specs/ChainSpec/SelectingBlockProducers.md

```diff
@@ -0,0 +1,335 @@
+
+### Steps
+
+```python
+# The seed is used as an entropy source for the random numbers.
+# The first 8 bytes select a block producer uniformly.
+uniform_index = int.from_bytes(rng_seed[0:8], byteorder='little') % len(validators)
+
+# The next 16 bytes uniformly pick some weight between 0 and the total
+# weight (i.e. stake) of all block producers.
+let uniform_weight = int.from_bytes(rng_seed[8:24], byteorder='little') \
+    % sampler.weight_sum()
+
+# Return either the uniformly selected block producer, or its "alias"
+# depending on the uniformly selected weight.
+index = uniform_index \
+    if uniform_weight < sampler.no_alias_odds[uniform_index] \
+    else sampler.aliases[uniform_index]
+
+return validators[index]
+```
+
+## Algorithm for selecting producer of block at height \\( h \\)
+
+### Input
+
+* `h: BlockHeight`
+  - Height to compute the block producer for
+  - Only heights within the epoch corresponding to the given block producers make sense as input
+* `block_producers: Vec<ValidatorStake>` (output from above)
+* `block_producer_sampler: WeightedIndex`
+* `epoch_rng_seed: [u8; 32]`
+  - Fixed seed for the epoch determined from Verified Random Function (VRF) output of last block in
+    the previous epoch
+
+### Output
+
+* `block_producer: ValidatorStake`
+
+### Steps
+
+```python
+# Concatenates the bytes of the epoch seed with the height,
+# then computes the sha256 hash.
+block_seed = combine(epoch_rng_seed, h)
+
+# Use the algorithm defined above
+return select_validator(rng_seed=block_seed, validators=block_producers, sampler=block_producer_sampler)
+```
+
+## Algorithm for selection of chunk producer at height \\( h \\) for all shards
+
+### Input
+
+* (same inputs as selection of block producer at height h)
+* `num_shards: usize`
+* `chunk_producer_sampler: Vec<WeightedIndex>` (outputs from chunk-only producer selection)
+* `validator_shard_assignments: Vec<Vec<ValidatorStake>>`
+
+### Output
+
+* `chunk_producers: Vec<ValidatorStake>`
+  - `i`th element gives the validator that will produce the chunk for shard `i`. Note: at least one
+    of these will be a block producer, while others will be chunk-only producers.
+
+### Steps
+
+```python
+bp = block_producer_at_height(
+    h + 1,
+    block_producers,
+    block_producer_sampler,
+    epoch_rng_seed,
+)
+
+result = []
+for shard_id in range(num_shards):
+    # concatenate bytes and take hash to create unique seed
+    shard_seed = combine(epoch_rng_seed, h, shard_id)
+    # Use selection algorithm defined above
+    cp = select_validator(
+        rng_seed=shard_seed,
+        validators=validator_shard_assignments[shard_id],
+        sampler=chunk_producer_sampler[shard_id]
+    )
+    result.append(cp)
+
+# Ensure the block producer for the next block also produces one of the shards.
+# `bp` could already be in the result because block producers are also
+# chunk producers (see algorithm for selecting chunk producers from proposals).
+if bp not in result:
```

Comment:
I am not sure whether this is a good idea. Block producers need to track all shards and only reducing the latency on one shard is not going to help a lot. Plus, this can result in some weird dynamics where a chunk only producer for some shard may not get an opportunity to produce any chunk, thereby losing their reward.

Reply:
Sure, I'm fine to drop this if we think it does more harm than good.